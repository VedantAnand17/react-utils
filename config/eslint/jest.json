{
  "env": {
    "browser": true,
    "jest": true
  },
  "plugins": [
    "jest"
  ],
  "rules": {
    "@typescript-eslint/no-implied-eval": 0,
    "global-require": 0,
    "import/no-dynamic-require": 0,
    "import/no-extraneous-dependencies": 0,
    "no-console": 0,
    "no-new-func": 0,
    "react/require-default-props": 0
  }
}