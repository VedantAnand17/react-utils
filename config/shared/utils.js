/**
 * Misc utilities reusable in different JS-format config files.
 */
/* eslint-disable import/no-extraneous-dependencies */

const fs = require('fs');
const path = require('path');
const cssesc = require('cssesc');

const { interpolateName } = require('loader-utils');

/**
 * Normalizes file path to OS-independent format (adopted from css-loader).
 * @ignore
 * @param {string} file
 * @return {string}
 */
function normalizePath(file) {
  return path.sep === '\\' ? file.replace(/\\/g, '/') : file;
}

// eslint-disable-next-line no-control-regex
const filenameReservedRegex = /[<>:"/\\|?*]/g;
// eslint-disable-next-line no-control-regex
const reControlChars = /[\u0000-\u001f\u0080-\u009f]/g;

function escapeLocalident(localident) {
  return cssesc(
    localident
      // For `[hash]` placeholder
      .replace(/^((-?[0-9])|--)/, '_$1')
      .replace(filenameReservedRegex, '-')
      .replace(reControlChars, '-')
      .replace(/\./g, '-'),
    { isIdentifier: true },
  );
}

/**
 * Returns the name of package containing the folder; i.e. it recursively looks
 * up from the folder for the closest package.json file, and returns the name in
 * that file. It also caches the results from previously fisited folders.
 * @ignore
 * @param {string} folder
 * @return {string}
 */
function getPackageInfo(folder) {
  let res = getPackageInfo.cache[folder];
  if (!res) {
    const p = path.resolve(folder, 'package.json');
    /* eslint-disable global-require, import/no-dynamic-require */
    res = fs.existsSync(p) ? {
      name: require(p).name,
      root: folder,
    } : getPackageInfo(path.resolve(folder, '..'));
    /* eslint-enable global-require, import/no-dynamic-require */
    getPackageInfo.cache[folder] = res;
  }
  return res;
}
getPackageInfo.cache = {};

function getLocalIdent(
  { resourcePath },
  localIdentName,
  localName,
  options = {},
) {
  const packageInfo = getPackageInfo(path.dirname(resourcePath));
  const request = normalizePath(path.relative(packageInfo.root, resourcePath));
  return interpolateName({
    resourcePath,
  }, localIdentName, {
    ...options,
    content: `${packageInfo.name + request}\x00${localName}`,
    context: packageInfo.root,
  }).replace(/\[package\]/gi, packageInfo.name)
    .replace(/\[local\]/gi, localName)
    .replace(/@/g, '-');
}

function generateScopedNameDev(localName, assetPath) {
  // Mind: "css-loader" additionally escapes identifiers returned by
  // getLocalIdent(); the babel plugins - don't, thus we need to escape
  // ourselves.
  return escapeLocalident(
    getLocalIdent(
      { resourcePath: assetPath },
      '[package]___[path][name]___[local]___[hash:base64:6]',
      localName,
      {},
    ),
  );
}

function generateScopedNameProd(localName, assetPath) {
  return escapeLocalident(
    getLocalIdent(
      { resourcePath: assetPath },
      '[hash:base64:6]',
      localName,
      {},
    ),
  );
}

module.exports = {
  generateScopedNameDev,
  generateScopedNameProd,
  getLocalIdent,
};
