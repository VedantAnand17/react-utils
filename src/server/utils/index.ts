// Aggregates utility modules together.

export * as errors from './errors';
