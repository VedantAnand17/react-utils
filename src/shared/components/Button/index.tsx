// The <Button> component implements a standard button / button-like link.

import { type ReactNode } from 'react';

import Link from 'components/Link';

import themed, { type Theme } from '@dr.pogodin/react-themes';

import defaultTheme from './style.scss';

type PropsT = {
  active?: boolean;
  children?: ReactNode;
  disabled?: boolean;
  enforceA?: boolean;
  onClick?: React.MouseEventHandler & React.KeyboardEventHandler;
  onMouseDown?: React.MouseEventHandler;
  openNewTab?: boolean;
  replace?: boolean;
  theme: Theme<'active' | 'button' | 'disabled'>;
  // TODO: It needs a more precise typing of the object option.
  to?: object | string;
};

/* eslint-disable react/function-component-definition */
export const BaseButton: React.FunctionComponent<PropsT> = ({
  active,
  children,
  disabled,
  enforceA,
  onClick,
  onMouseDown,
  openNewTab,
  replace,
  theme,
  to,
}) => {
  let className = theme.button;
  if (active && theme.active) className += ` ${theme.active}`;
  if (disabled) {
    if (theme.disabled) className += ` ${theme.disabled}`;
    return (
      <div className={className}>
        {children}
      </div>
    );
  }
  if (to) {
    return (
      <Link
        className={className}
        enforceA={enforceA}
        onClick={onClick}
        onMouseDown={onMouseDown}
        openNewTab={openNewTab}
        replace={replace}
        to={to}
      >
        {children}
      </Link>
    );
  }

  return (
    <div
      className={className}
      onClick={onClick}
      onKeyDown={onClick && ((e) => {
        if (e.key === 'Enter') onClick(e);
      })}
      onMouseDown={onMouseDown}
      role="button"
      tabIndex={0}
    >
      {children}
    </div>
  );
};

/**
 * Button component theme: a map of CSS
 * class names to append to button elements:
 * @prop {string} [active] to the root element of active button.
 * @prop {string} [button] to the root element of any button.
 * @prop {string} [disabled] to the root element of disabled button.
 */
export default themed(BaseButton, 'Button', defaultTheme);
