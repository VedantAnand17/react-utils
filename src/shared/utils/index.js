import path from 'path';

import config from './config';
import * as isomorphy from './isomorphy';
import * as redux from './redux';
import * as webpack from './webpack';

const juUrl = module.webpackPolyfill ? './shared/utils/jest'
  : path.resolve(__dirname, './jest');

const JU = isomorphy.isServerSide() ? webpack.requireWeak(juUrl) : null;

export {
  config,
  isomorphy,
  JU,
  redux,
  webpack,
};
