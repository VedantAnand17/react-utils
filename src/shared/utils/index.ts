import themedImpl, {
  type Theme,
  COMPOSE,
  PRIORITY,
  ThemeProvider,
} from '@dr.pogodin/react-themes';

import config from './config';
import * as isomorphy from './isomorphy';
import time from './time';
import * as webpack from './webpack';

import type * as JuT from './jest';

export {
  Barrier,
  Emitter,
  Semaphore,
  withRetries,
} from '@dr.pogodin/js-utils';

export { getSsrContext } from './globalState';
export { default as splitComponent } from './splitComponent';

type ThemedT = typeof themedImpl & {
  COMPOSE: typeof COMPOSE;
  PRIORITY: typeof PRIORITY;
};

const themed: ThemedT = themedImpl as ThemedT;

themed.COMPOSE = COMPOSE;
themed.PRIORITY = PRIORITY;

// Note: it should be done this way, as in some environments
// "process" might not exist, and process.env.NODE_CONFIG_ENV
// not injected by Webpack.
let NODE_CONFIG_ENV;
try {
  NODE_CONFIG_ENV = process.env.NODE_CONFIG_ENV;
} catch { /* noop */ }

const env = NODE_CONFIG_ENV || process.env.NODE_ENV;

const JU: typeof JuT | null = env !== 'production'
  ? webpack.requireWeak('./jest', __dirname) as typeof JuT | null
  : null;

export {
  type Theme,
  config,
  isomorphy,
  JU,
  themed,
  ThemeProvider,
  time,
  webpack,
};
