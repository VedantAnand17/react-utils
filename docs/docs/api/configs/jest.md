# Jest
Default configuration for [Jest] testing framework.

:::caution Beware
It does strong assumptions about the host project's code structure: the host
[Jest] configuration in the `/config/jest` folder of host code; the source code
in `/src` folder; build outputs in `/build` folder; _etc._

For a project with different structure it might be easier to compose
the appropriate Jest config from scratch.
:::

To use this default configuration create two files in the `/config/jest` folder
of host project:

```js title="/config/jest/config.js"
const config = require('@dr.pogodin/react-utils/config/jest/default');

// You can tune config as necessary here.

module.exports = config;
```

```js title="/config/jest/setup.js"
import '@dr.pogodin/react-utils/config/jest/setup';
```

After that point to that host config file from the Jest execution command,
_e.g._:
```bash
jest --config config/jest/config.js
```

[Jest]: https://jestjs.io
