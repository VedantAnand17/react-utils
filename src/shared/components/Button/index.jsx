/**
 * The <Button> component implements a standard button / button-like link.
 */

import PT from 'prop-types';
import React from 'react';

import Link from 'components/Link';

import { themed } from 'utils';

import defaultTheme from './style.scss';

function BaseButton({
  active,
  children,
  disabled,
  enforceA,
  onClick,
  onMouseDown,
  openNewTab,
  replace,
  theme,
  to,
}) {
  let className = theme.button;
  if (active && theme.active) className += ` ${theme.active}`;
  if (disabled) {
    if (theme.disabled) className += ` ${theme.disabled}`;
    return (
      <div className={className}>
        {children}
      </div>
    );
  }
  if (to) {
    return (
      <Link
        className={className}
        enforceA={enforceA}
        onClick={onClick}
        onMouseDown={onMouseDown}
        openNewTab={openNewTab}
        replace={replace}
        to={to}
      >
        {children}
      </Link>
    );
  }
  return (
    <div
      className={className}
      onClick={onClick}
      onKeyPress={onClick}
      onMouseDown={onMouseDown}
      role="button"
      tabIndex={0}
    >
      {children}
    </div>
  );
}

const ThemedButton = themed('Button', [
  'active',
  'button',
  'disabled',
], defaultTheme)(BaseButton);

BaseButton.defaultProps = {
  active: false,
  children: undefined,
  disabled: false,
  enforceA: false,
  onClick: undefined,
  onMouseDown: undefined,
  openNewTab: false,
  replace: false,
  to: undefined,
};

BaseButton.propTypes = {
  active: PT.bool,
  children: PT.node,
  disabled: PT.bool,
  enforceA: PT.bool,
  onClick: PT.func,
  onMouseDown: PT.func,
  openNewTab: PT.bool,
  replace: PT.bool,
  theme: ThemedButton.themeType.isRequired,
  to: PT.oneOfType([PT.object, PT.string]),
};

export default ThemedButton;

/* TODO: Deprecated 2020.05.02. Remove after some time. */
function DeprecatedBaseButton(...args) {
  console.warn('[react-utils]: <BaseButton /> export is deprecated');
  /* eslint-disable react/jsx-props-no-spreading */
  return <BaseButton {...args} />;
  /* eslint-disable react/jsx-props-no-spreading */
}
export { DeprecatedBaseButton as BaseButton };
