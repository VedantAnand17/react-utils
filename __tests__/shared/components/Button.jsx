import _ from 'lodash';
import Button from 'components/Button';
import React from 'react';
import { shallowSnapshot } from 'utils/jest';

const testTheme = {
  active: 'ACTIVE_CLASS',
  button: 'BUTTON_CLASS',
  disabled: 'DISABLED_CLASS',
};

describe('Matches snapshots', () => {
  test('when active', () => {
    shallowSnapshot((
      <Button
        active
        onClick={_.noop}
        theme={testTheme}
      >
        BUTTON
      </Button>
    ));
    shallowSnapshot((
      <Button
        active
        onClick={_.noop}
        theme={_.omit(testTheme, 'regular')}
      >
        BUTTON
      </Button>
    ));
  });
  test('when disabled', () => {
    shallowSnapshot((
      <Button
        disabled
        onClick={_.noop}
        theme={testTheme}
      >
        BUTTON
      </Button>
    ));
    shallowSnapshot((
      <Button
        disabled
        onClick={_.noop}
        theme={_.omit(testTheme, 'disabled')}
      >
        BUTTON
      </Button>
    ));
  });
  test('when rendered as link', () => {
    shallowSnapshot((
      <Button
        active
        theme={testTheme}
        to="/SOME/TEST/URL"
      >
        BUTTON
      </Button>
    ));
    shallowSnapshot((
      <Button
        active
        theme={_.omit(testTheme, 'link')}
        to="/SOME/TEST/URL"
      >
        BUTTON
      </Button>
    ));
  });
});
