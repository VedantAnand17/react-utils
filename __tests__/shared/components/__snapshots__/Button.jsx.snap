// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshots when active 1`] = `
<div
  className="-dr-pogodin-react-utils___src-shared-components-Button-style___button___E1FNQT   BUTTON_CLASS -dr-pogodin-react-utils___src-shared-components-Button-style___active___MAe9O6   ACTIVE_CLASS"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  BUTTON
</div>
`;

exports[`Matches snapshots when disabled 1`] = `
<div
  className="-dr-pogodin-react-utils___src-shared-components-Button-style___button___E1FNQT   BUTTON_CLASS -dr-pogodin-react-utils___src-shared-components-Button-style___disabled___Br9IWV   DISABLED_CLASS"
>
  BUTTON
</div>
`;

exports[`Matches snapshots when disabled 2`] = `
<div
  className="-dr-pogodin-react-utils___src-shared-components-Button-style___button___E1FNQT   BUTTON_CLASS -dr-pogodin-react-utils___src-shared-components-Button-style___disabled___Br9IWV"
>
  BUTTON
</div>
`;

exports[`Matches snapshots when rendered as link 1`] = `
<a
  className="-dr-pogodin-react-utils___src-shared-components-Button-style___button___E1FNQT   BUTTON_CLASS -dr-pogodin-react-utils___src-shared-components-Button-style___active___MAe9O6   ACTIVE_CLASS"
  disabled={false}
  href="/SOME/TEST/URL"
  onClick={[Function]}
  onMouseDown={null}
>
  BUTTON
</a>
`;
